src/hooks/useMaxSiblingWidth.ts

```diff
@@ -0,0 +1,144 @@
+  let i = currentIndex + direction
+
+  while (i >= 0 && i < siblings.length) {
+    const sibling = siblings[i]
+    const siblingDepth = sibling.getAttribute('data-depth')
+
+    if (sibling.classList.contains('tree-node') && siblingDepth === targetDepth) {
+      if (!sibling.querySelector('.divider')) {
+        collectedSiblings.push(sibling)
+      }
+    } else {
+      break
+    }
+
+    i += direction
+  }
+
+  return direction === -1 ? collectedSiblings.reverse() : collectedSiblings
+}
+
+/** Helper function to collect siblings at target depth and one level up. */
+const collectSiblingsOneLevelUp = (
+  siblings: HTMLElement[],
+  currentIndex: number,
+  targetDepth: string,
+): HTMLElement[] => {
+  const allowedDepth = String(parseInt(targetDepth, 10) - 1)
+  const collectedSiblings: HTMLElement[] = []
+
+  /** Collect siblings while moving in provided direction. */
+  const collect = (i: number, direction: -1 | 1) => {
+    while (i >= 0 && i < siblings.length) {
+      const sibling = siblings[i]
+      const siblingDepth = sibling.getAttribute('data-depth')
+
+      if (
+        sibling.classList.contains('tree-node') &&
+        siblingDepth &&
+        (siblingDepth === targetDepth || siblingDepth === allowedDepth)
+      ) {
+        if (!sibling.querySelector('.divider') && siblingDepth === targetDepth) {
+          collectedSiblings.push(sibling)
+        }
+      } else {
+        break
+      }
+
+      i += direction
+    }
+  }
+
+  collect(currentIndex - 1, -1)
+  collect(currentIndex + 1, 1)
+
+  return collectedSiblings
+}
+
+/** Helper function to calculate maximum width among '.thought' elements. */
+const calculateMaxWidth = (elements: HTMLElement[]): number => {
+  const widths = elements.map(el => {
+    const thought = el.querySelector<HTMLElement>('.thought')
+    return thought ? thought.offsetWidth : 0
+  })
+
+  return widths.length > 0 ? Math.max(...widths) : 0
+}
+
+/** Hook to calculate the width of a divider, equal to the max width of a sibling at a particular depth. */
+const useMaxSiblingWidth = (elRef: RefObject<HTMLElement>): number => {
+  const [maxWidth, setMaxWidth] = useState(0)
+
+  useEffect(() => {
+    const el = elRef.current
+    if (!el) return
+
+    // Get the closest parent node with the class .tree-node and its depth
+    const parentTreeNode = el.closest<HTMLElement>('.tree-node')
+    if (!parentTreeNode) return
+
+    const targetDepth = parentTreeNode.getAttribute('data-depth')
+    if (!targetDepth) return
+
+    // Get all siblings of the identified parent node
+    const siblings = Array.from(parentTreeNode.parentNode?.children || []) as HTMLElement[]
+
+    const currentIndex = siblings.indexOf(parentTreeNode)
+
+    // Collect consecutive siblings at the same depth
+    let consecutiveSiblings = [
+      ...collectConsecutiveSiblings(siblings, currentIndex, targetDepth, -1),
+      ...collectConsecutiveSiblings(siblings, currentIndex, targetDepth, 1),
```

Comment:
Can you explain the concept of `direction` that you are using? I'm not quite getting it. It appears that the collect functions are always called in both directions.

Is it possible to do this with a more straightforward map and filter?

Reply:
@raineorshine 

The problem is that the thought nodes including parents, children, grandchildren all at the same level in the DOM and positioned accordingly. There's no hierarchy.

So to find the sibling nodes, we need to traverse forward and backward until the depth of the node changes. That way we can collect all the sibling nodes.

map and filter cannot be used since we are trying to traverse from a point and only until a specific condition is met.

<img width="550" alt="image" src="https://github.com/cybersemics/em/assets/6959733/f38380ae-65e0-436d-97c4-37191b8041a0">

Reply:
I see, that makes sense now. Thank you for the thorough explanation and video.

So what you are doing is using the absolute depth to reconstruct the parent thought, to determine all of the thoughts with the same parent as the divider. However instead of encoding the depth in the DOM, why not encode the parent path? This can be used to uniquely identify all of the divider's siblings. We already have access to the parent path in React (i.e. `[roadmapId, hypertextId]`, accessed with `parentOf(path)`). You can encode it in the DOM and then query the matching parent path to determine the divider's siblings.

The same approach can be taken for the col2 only-child divider, you just need to store the grandparent path (i.e. the path of the table thought) to identify all col2 thoughts in the same table.